rust/src/storage/file.rs

```diff
@@ -58,26 +65,110 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
+        let tmp_file = self
+            .root
+            .join("_tmp_writes")
+            .join(Uuid::new_v4().to_string());
 
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        create_parent_dirs(&tmp_file).await?;
+        create_parent_dirs(Path::new(path)).await?;
 
+        let mut f = fs::File::create(&tmp_file).await?;
         f.write(obj_bytes).await?;
 
+        rename(tmp_file.to_str().unwrap(), path)?;
+
         Ok(())
     }
 }
+
+async fn create_parent_dirs(p: &Path) -> std::io::Result<()> {
+    if let Some(d) = p.parent() {
+        fs::create_dir_all(d).await?;
+    }
+    Ok(())
+}
+
+#[cfg(target_os = "linux")]
+const RENAME_NOREPLACE: c_uint = 1;
+
+#[cfg(target_os = "macos")]
+const RENAME_EXCL: c_uint = 4;
+
+extern "C" {
+    #[cfg(target_os = "linux")]
+    fn renameat2(
+        olddirfd: c_int,
+        oldpath: *const c_char,
+        newdirfd: c_int,
+        newpath: *const c_char,
+        flags: c_uint,
+    ) -> c_int;
+
+    #[cfg(target_os = "macos")]
+    fn renamex_np(from: *const c_char, to: *const c_char, flags: c_uint) -> c_int;
+}
+
+fn rename(from: &str, to: &str) -> Result<(), StorageError> {
+    let from_s = CString::new(from).unwrap();
+    let to_s = CString::new(to).unwrap();
+
+    unsafe {
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                if renameat2(0, from_s.as_ptr(), 0, to_s.as_ptr(), RENAME_NOREPLACE) == 0 {
+                   return Ok(())
+                }
```

Comment:
we should probably use something like https://crates.io/crates/errno to check for the error code and return a generic non-retryable error, otherwise it can get stuck in retry loop on stuff like permission error. I believe `EEXIST` is the error code that should be mapped to `StorageError::AlreadyExists`.

Reply:
Well, will test that with linux later, but in mac it's always `-1` for me. I feel like the safest way (at least for mac) is to test on `exists` call once `-1` is received so we won't get an inf loop

p.s. `-1` was received in mac for each case
- b already exists
- b has uncreated parent dirs
- a is missing

Reply:
ugh, didn't know that, from docs https://www.unix.com/man-page/mojave/2/renamex_np/
```
     A 0 value is returned if the operation succeeds, otherwise rename() returns -1 and the global variable errno indicates the reason for the failure
 ```    
yeap, https://crates.io/crates/errno worth a try, otherwise I figured how it works now

Reply:
addressed that in https://github.com/delta-io/delta-rs/pull/126/commits/c8dd3823fc338ef29a65245679c82a9ad45b0587

Reply:
also moved rename into separate module as it has grown big